### ISSUE TITLE
scp and crossload to the same host

### ISSUE THREAD

**Issue by unxed** [elfmz/far2l#1436]
Then both panels use the same SCP connection to same SCP host, why do we need to download and upload files then doing copy and move? Can't we just do "cp" and "mv" on a server side?

**Comment by elfmz** [comment]
1 Still only limited functionality possible - no visible progress etc
2 Too narrow usecase
3 Also you may just run far2l remotely there and you'll get this and more

**Comment by unxed** [comment]
Spent two evenings investigating how to build far2l for openwrt/ddwrt. They even have no glibc there, musl only. Also mips architecture. Cross compilation with also cross compiling half of deps with another c lib... Not an easy task.

About progress — may it be an option disabled by default?

**Comment by elfmz** [comment]
Building far2l for such minimalistic env is separate issue. Theoretically it must be buildable, may be with some missing plugins, like colorer (due to xerces dependency), anx wx of course is under big question, but other parts should work.

**Comment by unxed** [comment]
I'm familiar with Perl, will look into it

**Comment by unxed** [comment]
As for target architecture, the router I can test build on is big endian mips with uClibc:
system type             : Atheros AR9132 rev 2 (0xb9)
cpu model               : MIPS 24Kc V7.4

**Comment by unxed** [comment]
Do I understand right that for build process only "generator" mode is required?

**Comment by elfmz** [comment]
router is big endian? it can make some troubles :)

**Comment by unxed** [comment]
> router is big endian? it can make some troubles :)

Just googled "AR9132" a bit. Could not find data sheet. Found this:
https://github.com/Deoptim/atheros/blob/master/JTAG/AR9132.cfg:
```
target create $TARGETNAME mips_m4k -endian big -chain-position $TARGETNAME
```

**Comment by unxed** [comment]
As embedded systems are often single user, we can introduce embedded build mode without sudo releated stuff to save build time and storage space.

**Comment by elfmz** [comment]
i dont think sudo-related stuff consumes lot of space and build time

**Comment by unxed** [comment]
> tried to crosscompile under buildroot

Isn't buildroot a tool for generating firmware images, not individual software packages?

**Comment by elfmz** [comment]
yes, but IMHO its cross compilation as target is not same as host

**Comment by unxed** [comment]
looks like entire far2l.dir is recompiled now on each `make` run even if nothing changed inside

**Comment by unxed** [comment]
> yes, but IMHO its cross compilation as target is not same as host

I just don't really understand how to make build, for example, for a openwrt/ddwrt router. Okay, I can tell cmake to use a different compilation toolchain with a different target platform, but where do I get the necessary header files for different libraries, and how do I tell cmake to use them, not the system ones?

**Comment by elfmz** [comment]
need to google steps for it.. isnt it this: https://openwrt.org/docs/guide-developer/toolchain/crosscompile ?

**Comment by unxed** [comment]
> need to google steps for it.. isnt it this: https://openwrt.org/docs/guide-developer/toolchain/crosscompile ?

Unfortunately, this is nothing there about where to get headers for dependency libs and how to pass them to cmake. Some search gives me two cmake examples, an application and a library it depents on:
[Libubox](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/libs/libubox/Makefile;h=cc5d7f7ae6b6beaeb535030195252015a0a679ed;hb=HEAD)
[Ubox](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/system/ubox/Makefile;h=9ba4abdd581a1d54918d2574f8b1020a5662aed7;hb=HEAD)

But it still does not explain, how ubox build recipe knows where to get libubox header files. Ubox [CMakeLists.txt](https://git.openwrt.org/?p=project/ubox.git;a=blob;f=CMakeLists.txt;h=62de1d4b96480eff5e8ee9069ef9d8f130e982d9;hb=HEAD) has `INCLUDE_DIRECTORIES(/opt/local/include)`. But I still do not understand a way how library header files for target platform should appear at this path.

**Comment by elfmz** [comment]
From link above:
`If compilation aborts due to missing header files or shared objects, you might need to override CFLAGS and LDFLAGS to point to the staging_dir/target-architecture_uClibc-libcver/usr/include and …/usr/lib directories`

**Comment by unxed** [comment]
>  So first you have to compile full buildroot image for your device - it will build also toolchain, libc and then compiler and necessary headers will be useable

Yes, got it. But this is about system libs. And that about headers for libs that far2l depends on like `libuchardet-dev`?

**Comment by elfmz** [comment]
Yes, buildroot appeared to have xerces and many others (but no wx and no uchardet), dependencies are decribed in buildroot package files that i've just added (however they little bit drafts for now, need some refinement) - need just to add them to buildroot/packages directory, then reference added package from some category in packages/Config.in then run make menuconfig and enable there far2l - then make - and it will create full image with far2l and libs inside

**Comment by elfmz** [comment]
At least i was able add far2l into image i build for my Olinuxino Lime2 that i use as kind of smart home brains :)

**Comment by unxed** [comment]
You mean I should build dependencies in buildroot also? But many of them, like libssh, are already availble as openwrt packages, seems better to use them to save router space.

**Comment by elfmz** [comment]
Building them needed to get properly configured headers and libs, no need to copy them on target if they're already there

**Comment by unxed** [comment]
And libs abi is the same within minor versions, am i right?

**Comment by unxed** [comment]
При попытке собрать под buildroot некоторые проблемки с зависимостями colorer'а наблюдаются:
![изображение](https://user-images.githubusercontent.com/1151423/204869569-9e35a5f7-15f9-4fe9-9039-1ca0722afe56.png)
![изображение](https://user-images.githubusercontent.com/1151423/204869723-764a7b3f-207e-49b0-b94d-a0f32af4b304.png)

**Comment by unxed** [comment]
There is still two use cases for "crossload via cp/mv" feature:
1) router has too little flash memory to install far2l
2) router has ssh, but do not support installing any additional software

**Comment by unxed** [comment]
на shell такое совсем несложно должно делаться, по идее
